Asset referenced with CSS variables not found using application-builder

### Command

build, run

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

We are currently testing the new application builder on my client application. 

We are referencing quite often a SASS stylesheets which sets a CSS Custom Property to load the an asset (a picture). This stylesheet also contains mixins that's why we use it in our components stylesheets using the `use` keyword.

This works fine using webpack but I can't manage to make it work with the esbuild builder.

It seems that the compiler transforms the path into an incorrect one, it should not change as I stated that `assets` is an `externalDependency ` in the `angular.json` file.

You can check the repo for a minimal reproduction.

Let me know if I can help on this in any ways ! 

### Minimal Reproduction

https://github.com/wmarques/sass-issue-repro

npm install && npm start

### Exception or Error

```text
✘ [ERROR] Could not resolve "../styles/assets/img.png" [plugin angular-css-resource]

    src/app/app.component.scss:3:18:
      3 │   --img-path: url("../styles/assets/img.png");
        ╵                   ~~~~~~~~~~~~~~~~~~~~~~~~~~

  Preprocessor stylesheets may not show the exact file location of the error.


✘ [ERROR] Could not resolve "./styles/assets/img.png" [plugin angular-css-resource]

    src/styles.scss:3:18:
      3 │   --img-path: url("./styles/assets/img.png");
        ╵                   ~~~~~~~~~~~~~~~~~~~~~~~~~

  Preprocessor stylesheets may not show the exact file location of the error.
```


### Your Environment

```text
Angular CLI: 17.3.3
Node: 18.20.1
Package Manager: npm 10.5.0
OS: linux x64

Angular: 17.3.3
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1703.3
@angular-devkit/build-angular   17.3.3
@angular-devkit/core            17.3.3
@angular-devkit/schematics      17.3.3
@schematics/angular             17.3.3
rxjs                            7.8.1
typescript                      5.4.4
zone.js                         0.14.4
```


### Anything else relevant?

_No response_

--- a/packages/angular_devkit/build_angular/src/builders/application/tests/behavior/stylesheet-url-resolution_spec.ts
+++ b/packages/angular_devkit/build_angular/src/builders/application/tests/behavior/stylesheet-url-resolution_spec.ts
@@ -7,6 +7,7 @@
  */
 
 import { buildApplication } from '../../index';
+import { OutputHashing } from '../../schema';
 import { APPLICATION_BUILDER_INFO, BASE_OPTIONS, describeBuilder } from '../setup';
 
 describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
@@ -164,34 +165,18 @@ describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
       );
     });
 
-    it('should not rebase a URL with a namespaced Sass variable reference', async () => {
-      await harness.writeFile(
-        'src/styles.scss',
-        `
-        @import "a";
-      `,
-      );
-      await harness.writeFile(
-        'src/a.scss',
-        `
-        @use './b' as named;
-        .a {
-          background-image: url(named.$my-var)
-        }
-      `,
-      );
-      await harness.writeFile(
-        'src/b.scss',
-        `
-        @forward './c.scss' show $my-var;
-      `,
-      );
-      await harness.writeFile(
-        'src/c.scss',
-        `
-        $my-var: "https://example.com/example.png";
-      `,
-      );
+    it('should not rebase a URL with a namespaced Sass variable reference that points to an absolute asset', async () => {
+      await harness.writeFiles({
+        'src/styles.scss': `@use 'theme/a';`,
+        'src/theme/a.scss': `
+                            @use './b' as named;
+                            .a {
+                              background-image: url(named.$my-var)
+                            }
+        `,
+        'src/theme/b.scss': `@forward './c.scss' show $my-var;`,
+        'src/theme/c.scss': `$my-var: "https://example.com/example.png";`,
+      });
 
       harness.useTarget('build', {
         ...BASE_OPTIONS,
@@ -199,11 +184,112 @@ describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
       });
 
       const { result } = await harness.executeOnce();
-      expect(result?.success).toBe(true);
+      expect(result?.success).toBeTrue();
 
       harness
         .expectFile('dist/browser/styles.css')
         .content.toContain('url(https://example.com/example.png)');
+    });
+
+    it('should not rebase a URL with a Sass variable reference that points to an absolute asset', async () => {
+      await harness.writeFiles({
+        'src/styles.scss': `@use 'theme/a';`,
+        'src/theme/a.scss': `
+                            @import './b';
+                            .a {
+                              background-image: url($my-var)
+                            }
+        `,
+        'src/theme/b.scss': `$my-var: "https://example.com/example.png";`,
+      });
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        styles: ['src/styles.scss'],
+      });
+
+      const { result } = await harness.executeOnce();
+      expect(result?.success).toBeTrue();
+
+      harness
+        .expectFile('dist/browser/styles.css')
+        .content.toContain('url(https://example.com/example.png)');
+    });
+
+    it('should rebase a URL with a namespaced Sass variable referencing a local resource', async () => {
+      await harness.writeFiles({
+        'src/styles.scss': `@use 'theme/a';`,
+        'src/theme/a.scss': `
+                            @use './b' as named;
+                            .a {
+                              background-image: url(named.$my-var)
+                            }
+        `,
+        'src/theme/b.scss': `@forward './c.scss' show $my-var;`,
+        'src/theme/c.scss': `$my-var: "./images/logo.svg";`,
+        'src/theme/images/logo.svg': `<svg></svg>`,
+      });
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        outputHashing: OutputHashing.None,
+        styles: ['src/styles.scss'],
+      });
+
+      const { result } = await harness.executeOnce();
+      expect(result?.success).toBeTrue();
+
+      harness.expectFile('dist/browser/styles.css').content.toContain(`url("./media/logo.svg")`);
+      harness.expectFile('dist/browser/media/logo.svg').toExist();
+    });
+
+    it('should rebase a URL with a Sass variable referencing a local resource', async () => {
+      await harness.writeFiles({
+        'src/styles.scss': `@use 'theme/a';`,
+        'src/theme/a.scss': `
+                            @import './b';
+                            .a {
+                              background-image: url($my-var)
+                            }
+        `,
+        'src/theme/b.scss': `$my-var: "./images/logo.svg";`,
+        'src/theme/images/logo.svg': `<svg></svg>`,
+      });
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        outputHashing: OutputHashing.None,
+        styles: ['src/styles.scss'],
+      });
+
+      const { result } = await harness.executeOnce();
+      expect(result?.success).toBeTrue();
+
+      harness.expectFile('dist/browser/styles.css').content.toContain(`url("./media/logo.svg")`);
+      harness.expectFile('dist/browser/media/logo.svg').toExist();
+    });
+
+    it('should not process a URL that has been marked as external', async () => {
+      await harness.writeFiles({
+        'src/styles.scss': `@use 'theme/a';`,
+        'src/theme/a.scss': `
+          .a {
+            background-image: url("assets/logo.svg")
+          }
+        `,
+      });
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        outputHashing: OutputHashing.None,
+        externalDependencies: ['assets/*'],
+        styles: ['src/styles.scss'],
+      });
+
+      const { result } = await harness.executeOnce();
+      expect(result?.success).toBeTrue();
+
+      harness.expectFile('dist/browser/styles.css').content.toContain(`url(assets/logo.svg)`);
     });
   });
 });

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/css-resource-plugin.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/stylesheets/css-resource-plugin.ts
@@ -32,42 +32,51 @@ export function createCssResourcePlugin(cache?: LoadResultCache): Plugin {
     name: 'angular-css-resource',
     setup(build: PluginBuild): void {
       build.onResolve({ filter: /.*/ }, async (args) => {
+        const { importer, path, kind, resolveDir, namespace, pluginData = {} } = args;
+
         // Only attempt to resolve url tokens which only exist inside CSS.
         // Also, skip this plugin if already attempting to resolve the url-token.
-        if (args.kind !== 'url-token' || args.pluginData?.[CSS_RESOURCE_RESOLUTION]) {
+        if (kind !== 'url-token' || pluginData[CSS_RESOURCE_RESOLUTION]) {
           return null;
+        }
+
+        let [containingDir, resourceUrl] = path.split('||file:', 2);
+        if (resourceUrl === undefined) {
+          // This can happen due to early exit checks in rebasing-importer
+          // logic such as when the url is an external URL.
+          resourceUrl = containingDir;
+          containingDir = '';
         }
 
         // If root-relative, absolute or protocol relative url, mark as external to leave the
         // path/URL in place.
-        if (/^((?:\w+:)?\/\/|data:|chrome:|#|\/)/.test(args.path)) {
+        if (/^((?:\w+:)?\/\/|data:|chrome:|#|\/)/.test(resourceUrl)) {
           return {
-            path: args.path,
+            path: resourceUrl,
             external: true,
           };
         }
 
-        const { importer, kind, resolveDir, namespace, pluginData = {} } = args;
         pluginData[CSS_RESOURCE_RESOLUTION] = true;
 
-        const result = await build.resolve(args.path, {
+        const result = await build.resolve(resourceUrl, {
           importer,
           kind,
           namespace,
           pluginData,
-          resolveDir,
+          resolveDir: join(resolveDir, containingDir),
         });
 
         if (result.errors.length) {
           const error = result.errors[0];
-          if (args.path[0] === '~') {
+          if (resourceUrl[0] === '~') {
             error.notes = [
               {
                 location: null,
                 text: 'You can remove the tilde and use a relative path to reference it, which should remove this error.',
               },
             ];
-          } else if (args.path[0] === '^') {
+          } else if (resourceUrl[0] === '^') {
             error.notes = [
               {
                 location: null,

--- a/packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts
+++ b/packages/angular_devkit/build_angular/src/tools/sass/rebasing-importer.ts
@@ -82,18 +82,15 @@ abstract class UrlRebasingImporter implements Importer<'sync'> {
         continue;
       }
 
-      // Skip if value is a Sass variable.
-      // Sass variable usage either starts with a `$` or contains a namespace and a `.$`
-      if (value[0] === '$' || /^\w+\.\$/.test(value)) {
-        continue;
-      }
-
       // Skip if root-relative, absolute or protocol relative url
       if (/^((?:\w+:)?\/\/|data:|chrome:|#|\/)/.test(value)) {
         continue;
       }
 
-      const rebasedPath = relative(this.entryDirectory, join(stylesheetDirectory, value));
+      // Sass variable usage either starts with a `$` or contains a namespace and a `.$`
+      const valueNormalized = value[0] === '$' || /^\w+\.\$/.test(value) ? `#{${value}}` : value;
+      const rebasedPath =
+        relative(this.entryDirectory, stylesheetDirectory) + '||file:' + valueNormalized;
 
       // Normalize path separators and escape characters
       // https://developer.mozilla.org/en-US/docs/Web/CSS/url#syntax

